Sources/momento/internal/utils/Interceptors.swift

```diff
@@ -0,0 +1,264 @@
+    
+    private let apiKey: String
+    
+    init(apiKey: String) {
+        self.apiKey = apiKey
+    }
+    
+    func makePublishInterceptors() -> [ClientInterceptor<CacheClient_Pubsub__PublishRequest, CacheClient_Pubsub__Empty>] {
+        [AuthHeaderInterceptor(apiKey: apiKey)]
+    }
+
+    func makeSubscribeInterceptors() -> [ClientInterceptor<CacheClient_Pubsub__SubscriptionRequest, CacheClient_Pubsub__SubscriptionItem>] {
+        [AuthHeaderInterceptor(apiKey: apiKey)]
+    }
+}
+
+// Interceptor factory class for the CacheControlClient
+final class ControlClientInterceptorFactory: ControlClient_ScsControlClientInterceptorFactoryProtocol {
+    
+    private let apiKey: String
+    
+    init(apiKey: String) {
+        self.apiKey = apiKey
+    }
+    
+    func makeCreateCacheInterceptors() -> [GRPC.ClientInterceptor<ControlClient__CreateCacheRequest, ControlClient__CreateCacheResponse>] {
+        [AuthHeaderInterceptor(apiKey: apiKey)]
+    }
+    
+    func makeDeleteCacheInterceptors() -> [GRPC.ClientInterceptor<ControlClient__DeleteCacheRequest, ControlClient__DeleteCacheResponse>] {
+        [AuthHeaderInterceptor(apiKey: apiKey)]
+    }
+    
+    func makeListCachesInterceptors() -> [GRPC.ClientInterceptor<ControlClient__ListCachesRequest, ControlClient__ListCachesResponse>] {
+        [AuthHeaderInterceptor(apiKey: apiKey)]
+    }
+    
+    func makeFlushCacheInterceptors() -> [GRPC.ClientInterceptor<ControlClient__FlushCacheRequest, ControlClient__FlushCacheResponse>] {
+        [AuthHeaderInterceptor(apiKey: apiKey)]
+    }
+    
+    func makeCreateSigningKeyInterceptors() -> [GRPC.ClientInterceptor<ControlClient__CreateSigningKeyRequest, ControlClient__CreateSigningKeyResponse>] {
+        [AuthHeaderInterceptor(apiKey: apiKey)]
+    }
+    
+    func makeRevokeSigningKeyInterceptors() -> [GRPC.ClientInterceptor<ControlClient__RevokeSigningKeyRequest, ControlClient__RevokeSigningKeyResponse>] {
+        [AuthHeaderInterceptor(apiKey: apiKey)]
+    }
+    
+    func makeListSigningKeysInterceptors() -> [GRPC.ClientInterceptor<ControlClient__ListSigningKeysRequest, ControlClient__ListSigningKeysResponse>] {
+        [AuthHeaderInterceptor(apiKey: apiKey)]
+    }
+    
+    func makeCreateIndexInterceptors() -> [GRPC.ClientInterceptor<ControlClient__CreateIndexRequest, ControlClient__CreateIndexResponse>] {
+        [AuthHeaderInterceptor(apiKey: apiKey)]
+    }
+    
+    func makeDeleteIndexInterceptors() -> [GRPC.ClientInterceptor<ControlClient__DeleteIndexRequest, ControlClient__DeleteIndexResponse>] {
+        [AuthHeaderInterceptor(apiKey: apiKey)]
+    }
+    
+    func makeListIndexesInterceptors() -> [GRPC.ClientInterceptor<ControlClient__ListIndexesRequest, ControlClient__ListIndexesResponse>] {
+        [AuthHeaderInterceptor(apiKey: apiKey)]
+    }
+}
+
+// Interceptor factory class for the CacheDataClient
+final class DataClientInterceptorFactory: CacheClient_ScsClientInterceptorFactoryProtocol {
```

Comment:
this is very interesting, if I am reading this correctly, we are creating interceptors at the method level? So we have to define interceptors for every single method we define in the protocol

Reply:
yes, and it wouldn't build if I excluded the currently un-implemented functions, so I had to include them all and decided I may as well have each of them return the correct auth header interceptor

Reply:
Its an interesting approach swift took. On one hand, its super configurable and easy to define which methods get what interceptor, but its almost too configurable imo

Reply:
🤷

Reply:
🤷🏻‍♀️